### PR TITLE
eds: optionalize counting of unknown fields

### PR DESCRIPTION
**Pull request by pgenera**
Commit Message: In order to speed up eds, don't necessarily visit every proto field to count its validity as WarningValidationVisitor does. This yields a ~30% speed improvement in processing very large updates in EDS.
Risk Level: medium, new feature behind a command line flag.
Testing: Unit and bechmark tests.
Docs Changes: These are probably wrong, thus the draft-ness.
Release Notes: EDS can now ignore unknown dynamic fields, for a ~30% improvement in update processing time. Behind --ignore-unknown-dynamic-fields

### CODE REVIEW

File: docs/root/operations/admin.rst [hunk 1]
```diff
@@ -293,6 +293,7 @@ modify different aspects of the server:
         "service_zone": "",
         "mode": "Serve",
         "disable_hot_restart": false,
+	"disable_eds_validation": false,
```
Comment (htuch) [inline docs/root/operations/admin.rst:296]: Why is this control needed?

File: source/common/upstream/eds.cc [hunk 2]
```diff
@@ -117,9 +117,15 @@ void EdsClusterImpl::onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt
...
-      MessageUtil::anyConvertAndValidate<envoy::config::endpoint::v3::ClusterLoadAssignment>(
-          resources[0], validation_visitor_);
+  envoy::config::endpoint::v3::ClusterLoadAssignment cluster_load_assignment;
+  if (validation_visitor_.skipValidation()) {
+    cluster_load_assignment =
+        MessageUtil::anyConvert<envoy::config::endpoint::v3::ClusterLoadAssignment>(resources[0]);
+  } else {
+    cluster_load_assignment =
+        MessageUtil::anyConvertAndValidate<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+            resources[0], validation_visitor_);
```
Comment (htuch) [inline source/common/upstream/eds.cc:127]: Can we move this `skipValidation()` conditioning inside `MessageUtil::anyConvertAndValidate()`? This seems like something to do generically, rather than on a per-resource type basis, i.e. just for EDS here.

File: source/common/upstream/upstream_impl.cc [hunk 3]
```diff
@@ -1415,16 +1415,18 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
...
 
     if (existing_itr != existing_hosts_for_current_priority.end()) {
       existing_hosts_for_current_priority.erase(existing_itr);
-      itr = current_priority_hosts.erase(itr);
+      *itr = *(--erase_from_itr);
     } else {
       itr++;
     }
   }
+  current_priority_hosts.erase(erase_from_itr, current_priority_hosts.end());
```
Comment (htuch) [inline source/common/upstream/upstream_impl.cc:1429]: Is this a related change? If it's a general iterator safety thing or the like, can you split to  distinct PR?

File: test/common/upstream/eds_speed_test.cc [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  void doOnConfigUpdateVerifyNoThrow(
+      const envoy::config::endpoint::v3::ClusterLoadAssignment& cluster_load_assignment) {
+    Protobuf::RepeatedPtrField<ProtobufWkt::Any> resources;
+    resources.Add()->PackFrom(cluster_load_assignment);
+    VERBOSE_EXPECT_NO_THROW(eds_callbacks_->onConfigUpdate(resources, ""));
+  }
+
+  // Set up an EDS config with multiple priorities, localities, weights and make sure
+  // they are loaded and reloaded as expected.
+  void PriorityAndLocalityWeightedHelper(bool ignore_unknown_dynamic_fields, int num_hosts) {
```
Comment (htuch) [inline test/common/upstream/eds_speed_test.cc:114]: Please use Envoy style for naming, i.e. `priorityAndLocality..`

File: test/common/upstream/eds_speed_test.cc [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+          api_config_source:
+            api_type: REST
+            cluster_names:
+            - eds
+            refresh_delay: 1s
+    )EOF",
+                 Cluster::InitializePhase::Secondary);
+  }
+
+  void resetClusterDrainOnHostRemoval() {
```
Comment (htuch) [inline test/common/upstream/eds_speed_test.cc:53]: This seems pretty complicated for what you'd expect in a basic EDS performance benchmark to validate this change, how come the need for this kind of specialization?

File: test/common/upstream/eds_speed_test.cc (hunk 4 above)
Comment (pgenera) [reply test/common/upstream/eds_speed_test.cc]: Done.

File: source/common/upstream/upstream_impl.cc (hunk 3 above)
Comment (pgenera) [reply source/common/upstream/upstream_impl.cc:1429]: "related" insofar as it "makes it faster"; the original thought was removing this o(n^2) operation would make a difference. Testing showed it didn't make a big one, but I still think its worth doing. "unrelated" to the "provide a control to ignore unused fields" goal, however. I'm happy to pull it out if it feels risky.

File: source/common/upstream/eds.cc (hunk 2 above)
Comment (pgenera) [reply source/common/upstream/eds.cc:122]: Done.

File: docs/root/operations/admin.rst (hunk 1 above)
Comment (pgenera) [reply docs/root/operations/admin.rst]: Oops, its not!

File: test/common/upstream/eds_speed_test.cc (hunk 5 above)
Comment (pgenera) [reply test/common/upstream/eds_speed_test.cc:53]: There are competing goals of providing a lightweight speed test and measuring updates that look something like production EDS updates. That said I haven't given it much detailed consideration and can take a simplification pass. This test is mostly eds_test with large datasets wrapped in a benchmark test.

Comment (htuch) [reply test/common/upstream/eds_speed_test.cc:53]: I think from the perspective of what you want to achieve in this test, it would be better to simplify, since this is a microbenchmark.

File: include/envoy/server/options.h [hunk 6]
```diff
@@ -101,6 +101,11 @@ class Options {
    */
   virtual bool rejectUnknownDynamicFields() const PURE;
 
+  /**
+   * GENERA write the comment.
```
Comment (htuch) [inline include/envoy/server/options.h:105]: ?

File: include/envoy/protobuf/message_validator.h [hunk 7]
```diff
@@ -33,8 +33,16 @@ class ValidationVisitor {
    * @param description human readable description of the field
    */
   virtual void onUnknownField(absl::string_view description) PURE;
+
+  /**
+   * If true, skip this validation visitor in the interest of speed when
+   * possible.
+   **/
+  virtual bool skipValidation() { return false; }
```
Comment (htuch) [inline include/envoy/protobuf/message_validator.h:41]: We generally prefer pure virtual interfaces to remain pure, and push this default to the implementations.

File: source/common/upstream/upstream_impl.cc (hunk 3 above)
Comment (htuch) [reply source/common/upstream/upstream_impl.cc:1429]: Yeah, if you can split this into a later PR it would be great, thanks.

File: source/server/options_impl.cc [hunk 8]
```diff
@@ -81,6 +81,8 @@ OptionsImpl::OptionsImpl(std::vector<std::string> args,
   TCLAP::SwitchArg reject_unknown_dynamic_fields("", "reject-unknown-dynamic-fields",
                                                  "reject unknown fields in dynamic configuration",
                                                  cmd, false);
+  TCLAP::SwitchArg ignore_unknown_dynamic_fields("", "ignore-unknown-dynamic-fields",
+                                                 "GENERA more words.", cmd, false);
```
Comment (htuch) [inline source/server/options_impl.cc:85]: ?

File: test/common/upstream/eds_test.cc [hunk 9]
```diff
@@ -228,6 +228,17 @@ TEST_F(EdsTest, ValidateFail) {
   EXPECT_FALSE(initialized_);
 }
 
+// Validate that onConfigUpdate() can ignore unknown fields.
+TEST_F(EdsTest, ValidateIgnored) {
+  validation_visitor_.setSkipValidation(true);
+  initialize();
+  envoy::config::endpoint::v3::ClusterLoadAssignment resource;
+  Protobuf::RepeatedPtrField<ProtobufWkt::Any> resources;
+  resources.Add()->PackFrom(resource);
```
Comment (htuch) [inline test/common/upstream/eds_test.cc:237]: Where is the ignored field set?

File: test/common/upstream/eds_test.cc [hunk 10]
```diff
@@ -228,6 +228,17 @@ TEST_F(EdsTest, ValidateFail) {
...
+  // initialization fails but we don't throw, unlike ValidateFail above
+  EXPECT_FALSE(initialized_);
```
Comment (htuch) [inline test/common/upstream/eds_test.cc:239]: Should we verify that no stats were bumped or something like that?

File: test/common/upstream/eds_speed_test.cc (hunk 5 above)
Comment (pgenera) [reply test/common/upstream/eds_speed_test.cc]: I've streamlined this a bunch, but there's still some duplication between it and eds_test.

File: test/common/upstream/eds_test.cc (hunk 10 above)
Comment (pgenera) [reply test/common/upstream/eds_test.cc]: I've modified the integration test to validate that the counters aren't incremented.

File: source/server/options_impl.cc (hunk 8 above)
Comment (pgenera) [reply source/server/options_impl.cc:85]: I need to port my note-to-self presubmit to git, clearly ;)

Comment (htuch) [reply source/server/options_impl.cc:85]: Yeah, there are still a few of these in the PR, searching for "genera " in the diff.

File: test/common/upstream/eds_test.cc [hunk 11]
```diff
@@ -228,6 +228,24 @@ TEST_F(EdsTest, ValidateFail) {
...
 
+// Validate that onConfigUpdate() can ignore unknown fields.
+// this doesn't test the actual functionality, as the ValidationVisitor is mocked out,
+// however it is functionally tested in dynamic_validation_integration_test.
+TEST_F(EdsTest, ValidateIgnored) {
+  validation_visitor_.setSkipValidation(true);
+  initialize();
+  envoy::config::endpoint::v3::ClusterLoadAssignment resource;
+  resource.set_cluster_name("fare");
+  auto unknown = resource.GetReflection()->MutableUnknownFields(&resource);
```
Comment (htuch) [inline test/common/upstream/eds_test.cc:239]: Nit: `auto* unknown`

File: test/mocks/protobuf/mocks.h [hunk 12]
```diff
@@ -13,6 +13,13 @@ class MockValidationVisitor : public ValidationVisitor {
   ~MockValidationVisitor() override;
 
   MOCK_METHOD(void, onUnknownField, (absl::string_view));
+
+  bool skipValidation() override { return skip_validation_; }
+
+  void setSkipValidation(bool s) { skip_validation_ = s; }
```
Comment (htuch) [inline test/mocks/protobuf/mocks.h:19]: For mocks, we generally don't have setters and expose the fiels, but this doesn't hurt I guess.

File: test/integration/integration.h [hunk 13]
```diff
@@ -207,11 +207,12 @@ class BaseIntegrationTest : protected Logger::Loggable<Logger::Id::testing> {
   void createGeneratedApiTestServer(const std::string& bootstrap_path,
                                     const std::vector<std::string>& port_names,
                                     bool allow_unknown_static_fields,
-                                    bool reject_unknown_dynamic_fields, bool allow_lds_rejection);
+                                    bool reject_unknown_dynamic_fields,
+                                    bool ignore_unknown_dynamic_fields, bool allow_lds_rejection);
```
Comment (htuch) [inline test/integration/integration.h:211]: We now have 3 of these unknown field setting that are passed around, time for a `struct` wrapper?

File: test/integration/dynamic_validation_integration_test.cc [hunk 14]
```diff
@@ -46,33 +46,37 @@ class TestDynamicValidationNetworkFilterConfigFactory
 
 // Pretty-printing of parameterized test names.
 std::string dynamicValidationTestParamsToString(
-    const ::testing::TestParamInfo<std::tuple<Network::Address::IpVersion, bool>>& params) {
+    const ::testing::TestParamInfo<std::tuple<Network::Address::IpVersion, bool, bool>>& params) {
   return fmt::format(
-      "{}_{}",
+      "{}_{}_{}",
```
Comment (htuch) [inline test/integration/dynamic_validation_integration_test.cc:51]: How long does this test take to run now? Generally wary of uncontrolled combinatorial explosision. If it's not too long it's fine, but prefer to have only a meaningful subset of tests validated with this setting.

File: test/common/upstream/eds_speed_test.cc [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+#include "test/mocks/ssl/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/test_common/utility.h"
+
+#include "benchmark/benchmark.h"
+
+namespace Envoy {
+namespace Upstream {
+
+class EdsSpeedTest {
```
Comment (htuch) [inline test/common/upstream/eds_speed_test.cc:31]: This looks better now, +1.

File: test/integration/dynamic_validation_integration_test.cc (hunk 14 above)
Comment (pgenera) [reply test/integration/dynamic_validation_integration_test.cc:51]: 8.5s (locally on my corp machine). It's in the middle of the pack of //test/integration:all, with the longest ones in the 60s range.

File: test/integration/integration.h (hunk 13 above)
Comment (pgenera) [reply test/integration/integration.h]: Done, although I am less than perfectly confident the scoping is ideal.

File: source/server/options_impl.cc (hunk 8 above)
Comment (pgenera) [reply source/server/options_impl.cc]: I think I excised the last one (at least, grep -r thinks so too). I have not yet managed to make a local pre-commit that checks for them, but hey, something to hack on tomorrow.

File: docs/root/operations/cli.rst [hunk 16]
```diff
@@ -287,6 +287,11 @@ following are the command line options that Envoy supports.
   and these occurrences are counted in the :ref:`server.dynamic_unknown_fields <server_statistics>`
   statistic.
 
+.. option:: --ignore-unknown-dynamic-fields
+
+  *(optional)* This flag disables validation of protobuf configuration
+   for unknown fields in dynamic configuration, and ignores GENERA write more.
```
Comment (htuch) [inline docs/root/operations/cli.rst:293]: Nup, we still have some of these GENERA write mores! :)

Comment (pgenera) [reply docs/root/operations/cli.rst:293]: I think I'll choose to be glad past-pgenera was so thorough rather than be dismayed present-pgenera is situationally blind. :)

**Comment by htuch** [comment]
@pgenera see typo at https://dev.azure.com/cncf/envoy/_build/results?buildId=39159&view=logs&j=9fd74226-df5b-5f29-39da-76fc341cc763&t=068b98b1-acf0-52d0-fc6e-f6d1a5b1d5ef.

File: api/envoy/admin/v4alpha/server_info.proto [hunk 17]
```diff
@@ -96,6 +96,9 @@ message CommandLineOptions {
   // See :option:`--reject-unknown-dynamic-fields` for details.
   bool reject_unknown_dynamic_fields = 26;
 
+  // See :option:`--ignore_unknown_dynamic_fields` for details.
```
Comment (htuch) [inline api/envoy/admin/v4alpha/server_info.proto:99]: The docs build is broken because there is a `_` vs. `-` transposition here.

**Comment by pgenera** [comment]
The remaining clang-tidy complaint is pre-existing, I believe.
